### ISSUE TITLE
st_crs return NA if provided with EPSG code

### ISSUE THREAD

**Issue by felippehsk** [r-spatial/sf#2249]
Hi All,

Recently, I ran across an issue while running st_crs, when I stated the EPSG code and wanted to get the wkt format. I have been using sf for a long time and never had this issue before. As a matter of fact, I was using sf on the same machine and it was working. However, I had to update terra and R. Then, suddenly st_crs issues started to happen. 

**To Reproduce**
I am running R version 4.3.1 and sf 1.0-15. GDLA, GEOS and PROJ versions message is as follows:
`Linking to GEOS 3.11.2, GDAL 3.7.2, PROJ 9.3.0; sf_use_s2() is TRUE`

So when run the following code:
```
library(sf)
st_crs(4326)
```
I get the following response:
```
Coordinate Reference System: NA
Warning message:
In CPL_crs_from_input(x) :
  GDAL Error 1: PROJ: proj_create_from_database: C:/Program Files/R/R-4.3.1/library/rgdal/proj\proj.db contains DATABASE.LAYOUT.VERSION.MINOR = 0 whereas a number >= 2 is expected. It comes from another PROJ installation.
```

However, if I run:

```
fname <- system.file("shape/nc.shp", package="sf")
fname
## [1] "/home/runner/work/_temp/Library/sf/shape/nc.shp"
nc <- st_read(fname)
st_crs(nc)
```
The response is as expected:
```
Coordinate Reference System:
  User input: GCS_North_American_1927 
  wkt:
GEOGCRS["GCS_North_American_1927",
    DATUM["D_North_American_1927",
        ELLIPSOID["Clarke_1866",6378206.4,294.9786982,
            LENGTHUNIT["metre",1,
                ID["EPSG",9001]]]],
    PRIMEM["Greenwich",0,
        ANGLEUNIT["Degree",0.0174532925199433]],
    CS[ellipsoidal,2],
        AXIS["longitude",east,
            ORDER[1],
            ANGLEUNIT["Degree",0.0174532925199433]],
        AXIS["latitude",north,
            ORDER[2],
            ANGLEUNIT["Degree",0.0174532925199433]]]
```

**Additional context**
I have had the PROJ warning for a while now, it is something I still have to fix, but it was not a problem before.

**Comment by edzer** [comment]
Is your `PROJ_LIB` environment variable set? Does this still occur after removing `rgdal` (watch out: `rgdal` is no longer on CRAN).

**Comment by felippehsk** [comment]
Thank @edzer for the guidance!
I started by checking the path for `PROJ_LIB`. It was indeed using `rgdal` path. I removed `rgdal` and set the `PROJ_LIB` to OSGeo4W path for proj `C:\OSGeo4W\share\proj`. Now it works!

**Comment by rsbivand** [comment]
@felippehsk not a good idea, as the version of proj.db in OSGeo4W and Rtools43 may vary. For use in R on Windows, use the proj.db that is bundled with sf, unless you are sure that they are the same version.

**Comment by felippehsk** [comment]
@rsbivand, that is a great point. As soon as I removed `rgdal`, using `sf::sf_proj_search_paths()` it would show the proj.db bundled with `sf`. However, I would still get the same error as before. When using proj.db from OSGeo4W, I did not get the error. In this case, I would rephrase what I said before: It works as a temporary solution!

I am wondering here: what could be the cause of the issue I am experiencing? Any thoughts?

**Comment by edzer** [comment]
> what could be the cause of the issue I am experiencing?

The error message explains: the software and the proj.db version are out of sync. This should not happen if you have a binary install of `sf` from CRAN, and `PROJ_LIB` is not set to something external.

**Comment by rsbivand** [comment]
@felippehsk It is not clear which OS you are using, as OSGeo4W is Windows, `C:` also, but `/home/runner` could be a Linux/Unix/macOS/cloud instance. Are you using a single platform or multiple platforms? The version of `proj.db` can be inspected by:

```
library(RSQLite)
db <- dbConnect(SQLite(), dbname=<my proj.db>)
metadata <- dbReadTable(db, "metadata")
metadata[c(1:4, 13:14),]
dbDisconnect(db)
```
Changes in `DATABASE.LAYOUT.VERSION` will mean that the lookup in PROJ may not find what it is asked to look for, that is the version of PROJ itself linked to GDAL and `sf`and the `PROJ.VERSION` used to build `proj.db` need to be the same.

**Comment by felippehsk** [comment]
@edzer, yes! The error itself I understood, but after removing `rgdal` and re-installing `sf` from CRAN, I would still get the error. Even though `PROJ_LIB` was set to the `proj.db` under `sf`. I checked it by running `sf::sf_proj_search_paths()`. I also made sure to change the environmental variable in the system, but no success! I still got the same error. That is when I decided to try using OSGeo4W `proj.db`, which worked. 

@rsbivand, I currently running it on a Windows Server. I will try inspecting the database and bring more information to this discussion.

**Comment by edzer** [comment]
I would suggest completely removing package `sf`, then reinstalling, then retry. No clue how one could debug this sensibly.

**Comment by felippehsk** [comment]
@edzer, thank you again for the help. I understand this is hard to debug. I still didn't have time to look at the `proj.db` under `sf` install. Perhaps that could help track this down and find the reason for this issue. I would like to do that before trying to remove and re-installing `sf` again. BTW, after removing `rgdal`, I completely removed `sf` and installed it again. However, it did not fix the issue I was experiencing. 

If we all agree, this issue can be closed, and once I gather more information we can reopen it.
